os/arch/arm/src/amebad/amebad_timer_lowerhalf.c

```diff
@@ -0,0 +1,348 @@
+	return -ENODEV;
+}
+
+/****************************************************************************
+ * Name: amebad_gpt_settimeout
+ *
+ * Description:
+ *   Set a new timeout value (and reset the timer)
+ *
+ * Input Parameters:
+ *   lower   - A pointer the publicly visible representation of the
+ *             "lower-half" driver state structure.
+ *   timeout - The new timeout value in microseconds.
+ *
+ * Returned Values:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+static int amebad_gpt_settimeout(struct timer_lowerhalf_s *lower, uint32_t timeout)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+
+	if (priv->started) {
+		return -EPERM;
+	}
+
+	gtimer_reload(&priv->obj, timeout);
+	priv->gpt_timeout = timeout;
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: amebad_gpt_setcallback
+ *
+ * Description:
+ *   Call this user provided timeout callback.
+ *
+ * Input Parameters:
+ *   lower    - A pointer the publicly visible representation of the
+ *              "lower-half" driver state structure.
+ *   callback - The new timer expiration function pointer. If this
+ *              function pointer is NULL, then the reset-on-expiration
+ *              behavior is restored,
+ *  arg       - Argument that will be provided in the callback
+ *
+ * Returned Values:
+ *   The previous timer expiration function pointer or NULL is there was
+ *   no previous function pointer.
+ *
+ ****************************************************************************/
+static void amebad_gpt_setcallback(struct timer_lowerhalf_s *lower, tccb_t callback, void *arg)
+{
+	struct amebad_gpt_lowerhalf_s *priv = (struct amebad_gpt_lowerhalf_s *)lower;
+	irqstate_t flags = irqsave();
+
+	/* Save the new callback */
+	priv->callback = callback;
+	priv->arg = arg;
+
+	/* If no callback is attached, the timer stop at the first interrupt */
+	if (callback != NULL && priv->started) {
+		priv->obj.handler = (void *)amebad_timout_handler;
+		gtimer_init(&priv->obj, priv->obj.timer_id);
+	} else {
+		gtimer_deinit(&priv->obj);
+	}
+
+	irqrestore(flags);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+int amebad_timer_initialize(const char *devpath, int timer)
+{
+	struct amebad_gpt_lowerhalf_s *priv = NULL;
+
+	switch (timer) {
+	case TIMER0:
+		priv = &g_gpt0_lowerhalf;
+		priv->obj.timer_id = TIMER0;
+		break;
+
+	case TIMER1:
+		priv = &g_gpt1_lowerhalf;
+		priv->obj.timer_id = TIMER1;
+		break;
+
+	case TIMER2:
+		printf("Timer2\n");
+		priv = &g_gpt2_lowerhalf;
+		priv->obj.timer_id = TIMER2;
+		break;
+
+	case TIMER3:
+		printf("Timer3\n");
+		priv = &g_gpt3_lowerhalf;
+		priv->obj.timer_id = TIMER3;
+		break;
+	}
+
+	if (!priv) {
```

Comment:
I think this never comes..

Reply:
You are correct, I will add a default case for that.

Reply:
without this state, you can return -ENODEV in default.

Reply:
Sure, thanks!